crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs

```diff
@@ -0,0 +1,972 @@
+    arcs.iter()
+        .enumerate()
+        // For each agent's arc,
+        .map(|(i, arc_i)| {
+            // find the union of all arcs *other* than this one,
+            let other_arcset = arcs
+                .iter()
+                .enumerate()
+                .filter_map(|(j, arc_j)| {
+                    if i == j {
+                        None
+                    } else {
+                        Some(DhtArcSet::from(arc_j))
+                    }
+                })
+                .fold(DhtArcSet::new_empty(), |a, b| DhtArcSet::union(&a, &b));
+            // and return the intersection of this arc with the union of the others.
+            DhtArcSet::from(arc_i).intersection(&other_arcset)
+        })
+        // and take the union of all of the intersections
+        .fold(DhtArcSet::new_empty(), |a, b| DhtArcSet::union(&a, &b))
+}
+
+impl RoundState {
+    fn increment_sent_ops_blooms(&mut self) -> u8 {
+        self.num_sent_ops_blooms += 1;
+        self.num_sent_ops_blooms
+    }
+}
+
+/// Time range from now into the past.
+/// Start must be < end.
+fn time_range(start: Duration, end: Duration) -> Range<u64> {
+    let now = SystemTime::now();
+    let start = now
+        .checked_sub(start)
+        .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
+        .map(|t| t.as_millis() as u64)
+        .unwrap_or(0);
+
+    let end = now
+        .checked_sub(end)
+        .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
+        .map(|t| t.as_millis() as u64)
+        .unwrap_or(0);
+
+    start..end
+}
+
+kitsune_p2p_types::write_codec_enum! {
+    /// SimpleBloom Gossip Wire Protocol Codec
+    codec ShardedGossipWire {
+        /// Initiate a round of gossip with a remote node
+        Initiate(0x10) {
+            /// The list of arc intervals (equivalent to a [`DhtArcSet`])
+            /// for all local agents
+            intervals.0: Vec<ArcInterval>,
+            /// A random number to resolve concurrent initiates.
+            id.1: u8,
+        },
+
+        /// Accept an incoming round of gossip from a remote node
+        Accept(0x20) {
+            /// The list of arc intervals (equivalent to a [`DhtArcSet`])
+            /// for all local agents
+            intervals.0: Vec<ArcInterval>,
+        },
+
+        /// Send Agent Info Boom
+        Agents(0x30) {
+            /// The bloom filter for agent data
+            filter.0: PoolBuf,
+        },
+
+        /// Any agents that were missing from the remote bloom.
+        MissingAgents(0x40) {
+            /// The missing agents
+            agents.0: Vec<Arc<AgentInfoSigned>>,
+        },
+
+        /// Send Agent Info Boom
+        Ops(0x50) {
+            /// The bloom filter for op data
+            // maackle: what is the meaning and purpose of this ever being None?
+            //          currently looks like it means the same thing as
+            //          `finished == true`
+            filter.0: Option<PoolBuf>,
```

Comment:
Is there a reason the timing params are encoded opaquely in here? I think it might be more understandable if you had

```rust
time_window_start.1: u64,
time_window_end.2: u64,
```

or if you want them encoded in the option, perhaps:

```rust
filter.0: Option<(PoolBuf, u64, u64)>,
```